db/change_cache.go

```diff
@@ -577,31 +577,118 @@ func (c *changeCache) releaseUnusedSequence(ctx context.Context, sequence uint64
 	}
 }
 
-// releaseUnusedSequenceRange calls processEntry for each sequence in the range, but only issues a single notify.
+// releaseUnusedSequenceRange will handle unused sequence range arriving over DCP. It will batch remove from skipped or
+// push a range to pending sequences, or both.
 func (c *changeCache) releaseUnusedSequenceRange(ctx context.Context, fromSequence uint64, toSequence uint64, timeReceived time.Time) {
 
 	base.InfofCtx(ctx, base.KeyCache, "Received #%d-#%d (unused sequence range)", fromSequence, toSequence)
 
 	unusedSeq := channels.NewID(unusedSeqKey, unusedSeqCollectionID)
 	allChangedChannels := channels.SetOfNoValidate(unusedSeq)
-	for sequence := fromSequence; sequence <= toSequence; sequence++ {
+
+	// if range is single value, just run sequence through process entry and return early
+	if fromSequence == toSequence {
 		change := &LogEntry{
-			Sequence:     sequence,
+			Sequence:     toSequence,
 			TimeReceived: timeReceived,
 		}
-
-		// Since processEntry may unblock pending sequences, if there were any changed channels we need
-		// to notify any change listeners that are working changes feeds for these channels
 		changedChannels := c.processEntry(ctx, change)
 		allChangedChannels = allChangedChannels.Update(changedChannels)
 		c.channelCache.AddUnusedSequence(change)
+		if c.notifyChange != nil {
+			c.notifyChange(ctx, allChangedChannels)
+		}
+		return
 	}
 
+	// acquire mutex on cache for reading/writing to pending + reading next sequence value
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if toSequence < c.nextSequence {
+		// batch remove from skipped
+		err := c.RemoveSkippedSequences(ctx, fromSequence, toSequence)
+		if err != nil {
+			base.DebugfCtx(ctx, base.KeyCache, err.Error())
+			// we have sequences (possibly duplicate sequences) in the unused range that don't exist in skipped list
+			seqRange := &channels.UnusedSequenceRange{
```

Comment:
Is there benefit in building an UnusedSequenceRange, as opposed to just having startSeq and endSeq parameters on removeSubsetOfRangeFromSkipped?  It feels like we're cluttering up the code building these structs without any functional/performance benefit.